File: 0907-koko-eating-bananas/0907-koko-eating-bananas.py
class Solution:
    def isPossible(self, piles: List[int], h: int, mid: int) -> bool:
        hourNeeded = 0
        for pile in piles:
            hourNeeded += (pile + mid - 1) // mid  
        return hourNeeded <= h

    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        st = 1
        end = max(piles)  
        ans = 0
        
        while st <= end:
            mid = (st + end) // 2
            if self.isPossible(piles, h, mid):
                ans = mid
                end = mid - 1
            else:
                st = mid + 1

        return ans

File: 0907-koko-eating-bananas/0907-koko-eating-bananas.cpp
class Solution {
public:

    bool isPossible(vector<int>&piles, int n, int h, int mid)
    {
        long long hourNeeded = 0;
        for(int i = 0; i<n; i++)
        {
            hourNeeded += (piles[i] + mid - 1) / mid;
        }
        return hourNeeded <= h;
    }

    int minEatingSpeed(vector<int>& piles, int h) {
        int n = piles.size();
        long long st = 1;
        long long end = *max_element(piles.begin(), piles.end());
        int ans = 0;
        while(st <= end)
        {
            long long mid = st + (end - st )/ 2;
            if(isPossible(piles, n, h, mid))
            {
                ans = mid;
                end = mid - 1;
            }else
            {
                st = mid + 1;
            }
        }
        return ans;
    }
};